### PR TITLE
Source generator part 10: Add OutputGraphTypeGenerator

### PR DESCRIPTION
**Pull request by Shane32**
### Summary
This PR adds a new source generator for creating AOT-compatible output graph types (objects and interfaces) in GraphQL.NET. The generator produces optimized code for scenarios where reflection-based type creation is not available or desired.

#### New Generator
- **[`OutputGraphTypeGenerator.cs`](src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs)**: Core generator class that produces auto-registering graph type code
  - Supports both object and interface graph types
  - Handles nested partial class hierarchies
  - Generates field resolvers for properties, fields, and methods
  - Supports multiple instance creation strategies:
    - `ContextSource`: Uses context.Source directly
    - `GetRequiredService`: Resolves from DI container
    - `NewInstance`: Creates new instance with constructor parameters
    - `GetServiceOrCreateInstance`: Tries DI first, then falls back to new instance
  - Handles constructor parameters and required properties (C# 11+)
  - Supports static members
  - Special handling for subscription source stream resolvers

#### Key Features
- **Member Resolution**: Generates resolvers for properties, fields, and methods with parameters
- **Type Safety**: Uses fully qualified type names throughout generated code
- **DI Integration**: Supports service resolution via `BuildConstructorParameter<T>()`
- **Subscriptions**: Properly handles `IObservable<T>`, `Task<IObservable<T>>`, and `IAsyncEnumerable<T>` as source stream resolvers
- **Interface Types**: Correctly generates interface graph types with null resolvers for instance members

#### Tests
- **[`OutputGraphTypeGeneratorTests.cs`](src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.cs)**: Comprehensive unit tests covering:
  - Interface and object type generation
  - Empty types
  - Nested partial class hierarchies
  - All instance creation strategies
  - Constructor parameters and required properties
  - Subscription types with source stream resolvers
  - Edge cases (null data, empty hierarchy)

- **8 approval test files** verifying generated code output for different scenarios

### Technical Details

The generator produces code like:
- Dictionary-based member registration for efficient lookup
- Inline resolver generation to avoid reflection at runtime
- Proper handling of method parameter binding
- Support for both static and instance members
- Clean separation between interface and object type generation

### Impact
- Enables AOT compilation scenarios for GraphQL.NET
- Improves runtime performance by eliminating reflection
- Provides foundation for source-generated schema definitions

## See also
- #4338 
- #4350 
- #4355

### CODE REVIEW

File: src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+        // Build resolver - different for source stream resolvers
+        if (member.IsSourceStreamResolver)
+        {
+            sb.AppendLine($"{indent}fieldType.Resolver = global::GraphQL.Resolvers.SourceFieldResolver.Instance;");
+            sb.AppendLine($"{indent}fieldType.StreamResolver = BuildSourceStreamResolver(context =>");
+        }
+        else
+        {
+            sb.AppendLine($"{indent}fieldType.Resolver = BuildFieldResolver(context =>");
```
Comment (Shane32) [inline src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs:484]: BuildFieldResolver has multiple overloads to support synchronous or asynchronous resolvers, so the source generator does not need conditionals based on the return type.  BuildSourceStreamResolver has overloads for both `IObservable` and `IAsyncEnumerable`, both synchronous and asynchronous of each.  So the source generator fully supports all the types of methods that GraphQL.NET is currently capable of handling.

File: src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+        if (member.IsStatic)
+        {
+            // Static property/field access - use fully qualified type name
+            var declaringType = member.DeclaringTypeFullyQualifiedName ?? outputGraphType.FullyQualifiedClrTypeName;
+            sb.AppendLine($"{indent}fieldType.Resolver = BuildFieldResolver(context => {declaringType}.{member.MemberName}, false);");
+        }
+        else
+        {
+            // Instance property/field access
+            sb.AppendLine($"{indent}fieldType.Resolver = BuildFieldResolver(context => GetMemberInstance(context).{member.MemberName}, false);");
```
Comment (Shane32) [inline src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs:557]: The `, false` here signifies that the resolve field context accessor does not need to be set for this resolver, mimicking the same behavior as usual for a minor performance benefit.  Methods (see line 536) enable the accessor, as usual.

File: src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+        {
+            var member = outputGraphType.SelectedMembers[i];
+
+            if (!member.IsStatic && outputGraphType.IsInterface)
+            {
+                // Skip instance members for interfaces
+                continue;
+            }
+
+            sb.AppendLine($"{indent}public void ConstructField{i}_{member.MemberName}(global::GraphQL.Types.FieldType fieldType, global::System.Reflection.MemberInfo memberInfo)");
```
Comment (Shane32) [inline src/GraphQL.Analyzers.SourceGenerators/Generators/OutputGraphTypeGenerator.cs:438]: I guarantee unique method names by using the index, and to be user-friendly, I append the member name as well.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        private readonly global::System.Collections.Generic.Dictionary<global::System.Reflection.MemberInfo, global::System.Action<global::GraphQL.Types.FieldType, global::System.Reflection.MemberInfo>?> _members;
+        public AutoOutputGraphType_IStarWarsCharacter()
+        {
+            _members = new()
+            {
+                { typeof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter).GetProperty(nameof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.Id))!, null },
+                { typeof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter).GetProperty(nameof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.Name))!, null },
+                { typeof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter).GetMethod(nameof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.GetFriends), [typeof(global::GraphQL.StarWars.TypeFirst.StarWarsData)])!, null },
+                { typeof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter).GetMethod(nameof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.GetFriendsConnection), [typeof(global::GraphQL.StarWars.TypeFirst.StarWarsData), typeof(int?)])!, null },
```
Comment (Shane32) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:18]: GetMethod is necessary for the existing auto-registering logic to be functional.  It's also trimming-safe.  And we 'fully-qualify' the method name by specifying each of the parameter types as well.  Note that the first part (`typeof(...)`) may change for interfaces where the interface implements another interface; the specific interface for the method that is picked up by the type scanner is specified there.

This should all mimic the behavior in the base library.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+public partial class SampleAotSchema
+{
+    private class AutoOutputGraphType_IStarWarsCharacter : global::GraphQL.Types.Aot.AotAutoRegisteringInterfaceGraphType<global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter>
+    {
+        private readonly global::System.Collections.Generic.Dictionary<global::System.Reflection.MemberInfo, global::System.Action<global::GraphQL.Types.FieldType, global::System.Reflection.MemberInfo>?> _members;
+        public AutoOutputGraphType_IStarWarsCharacter()
+        {
+            _members = new()
+            {
+                { typeof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter).GetProperty(nameof(global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.Id))!, null },
```
Comment (Shane32) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:15]: This is an interface so most of the `Action<>` values are null.  This is because the only reason to support a resolver on an interface definition is for federation, and those are static resolvers.  So only static resolvers are generated.  Even then, if `!field.IsPrivate` they get cleared by GraphQL.NET.  This is exactly the same behavior as the base library.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            if (!fieldType.IsPrivate)
+            {
+                fieldType.Resolver = null;
+                fieldType.StreamResolver = null;
+            }
+        }
+
+        public void ConstructField5_Version(global::GraphQL.Types.FieldType fieldType, global::System.Reflection.MemberInfo memberInfo)
+        {
+            fieldType.Resolver = BuildFieldResolver(context => global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter.Version, false);
```
Comment (Shane32) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:42]: `Version` is a static property just for testing purposes only here btw.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.cs [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+/*
+ * 
+ * These tests do not rely on other components
+ * 
+ */
+
+/// <summary>
+/// Unit tests for OutputGraphTypeGenerator that generates AOT output graph type code.
+/// </summary>
+public class OutputGraphTypeGeneratorTests
```
Comment (Shane32) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.cs:15]: These tests supply a set of inputs to the source generator and verify the output.  The output is not compiled or verified for correct syntax.  This is unlike most of the prior tests which use a 'fake' source generator to process the input, but it does segregate the tests so it is not reliant on all the prior components.  The final 'part 12' of the source generator will include end-to-end tests that will demonstrate the code compiles correctly.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesGetServiceOrCreateInstance.WithRequiredProperties.approved.txt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        private readonly global::System.Collections.Generic.Dictionary<global::System.Reflection.MemberInfo, global::System.Action<global::GraphQL.Types.FieldType, global::System.Reflection.MemberInfo>> _members;
+        private readonly global::System.Func<global::GraphQL.IResolveFieldContext, global::Sample.Person> _getMemberInstance;
+        public AutoOutputGraphType_Person()
+        {
+            var ctorParam0 = BuildConstructorParameter<global::System.String>();
+            var ctorParam1 = BuildConstructorParameter<global::System.Int32>();
+            var reqProp0 = BuildConstructorParameter<global::System.String>();
+            var reqProp1 = BuildConstructorParameter<global::System.Int32>();
+            _getMemberInstance = context =>
```
Comment (Shane32) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesGetServiceOrCreateInstance.WithRequiredProperties.approved.txt:18]: Note that this is a very poor sample of how you'd actually use GetServiceOrCreateInstance, but it is generating the code correctly, pulling `string` and `int` from the service provider to inject into the ctor and required properties.

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt [hunk 9]
```diff
@@ -0,0 +1,54 @@
+// <auto-generated/>
+#nullable enable
+
+namespace AotSample;
+
+public partial class SampleAotSchema
+{
+    private class AutoOutputGraphType_IStarWarsCharacter : global::GraphQL.Types.Aot.AotAutoRegisteringInterfaceGraphType<global::GraphQL.StarWars.TypeFirst.IStarWarsCharacter>
+    {
+        private readonly global::System.Collections.Generic.Dictionary<global::System.Reflection.MemberInfo, global::System.Action<global::GraphQL.Types.FieldType, global::System.Reflection.MemberInfo>?> _members;
```
Comment (gao-artur) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:10]: Consider using FrozenDictionary

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        public AutoOutputGraphType_IStarWarsCharacter()
+        {
+            _members = new()
```
Comment (gao-artur) [inline src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:13]: It looks like the only thing preventing `_member` from being `static` is the usage of virtual methods defined on the base type. But do they have to be virtual? These base types are intended only for use by the source generator, right? If so, the user won't be able to override these methods anyway, and you can create additional methods with different behavior if needed. I think making `_member` static can boost the performance for non-singleton graphs. Then their type can be safely changed to `FrozenDictionary` without risking performance degradation due to frequent recreation.

Comment (gao-artur) [reply src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:13]: Except that `BuildArgument` can't be easily converted to static...

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesGetServiceOrCreateInstance.WithRequiredProperties.approved.txt (hunk 8 above)
Comment (gao-artur) [reply src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesGetServiceOrCreateInstance.WithRequiredProperties.approved.txt:18]: Oh, that's just an example. I scratched my head when I saw this in the previous file and tried to understand what I was missing 😄

File: src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt (hunk 10 above)
Comment (Shane32) [reply src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:13]: `BuildFieldResolver` and `BuildSourceStreamResolver` could easily be made static.  `BuildArgument` could be made static by modifying all its calls to use the static helpers directly (it would prevent those helpers from being overridden in source-generated code, but it should not matter).  The last remaining piece is `GetMemberInstance`.  Right now there's a default implementation that's overridden if needed.  When overridden, the source-generated code relies on `BuildConstructorParameter`, which also could be made static.  So `GetMemberInstance` could also be made static, although it would need to always be defined (as it couldn't be overridden).  

In the end, does it really provide a notable benefit?  We've never tried to optimize too much for scoped schemas in the past.  I can do it if you think we should; it will provide some benefit.  Keep in mind that it still will use reflection to scan for attributes and apply them to fields and arguments.  It also still uses a bunch of reflection to assign the graph type for each field.  The few assignments to `_members` is probably the least of the problems.

It would probably be better to forget that and allow the AOT graphs to [benefit from caching](https://github.com/graphql-dotnet/graphql-dotnet/pull/3728), like the current auto-registering types do.  This eliminates ALL reflection by simply copying the graph and field configuration from a cached instance.

As for `FrozenDictionary`, I'm not interested in using it, mainly since it's not available for .NET Standard 2.0 so I'd have to add code to check the target framework.  Further, [tests show that it's slower for very small dictionaries](https://code-corner.dev/2023/11/08/NET-8-%E2%80%94-FrozenDictionary-performance/), which these will often be.  So a plain dictionary should be faster (or comparable) for most graphs.

Comment (Shane32) [reply src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:13]: I added a todo entry in #4338  for supporting caching by AOT graph types.

Comment (gao-artur) [reply src/GraphQL.Analyzers.Tests/SourceGenerators/Generators/OutputGraphTypeGeneratorTests.GeneratesComprehensiveOutputGraphTypes.Interface.approved.txt:13]: > The few assignments to _members is probably the least of the problems.

It makes allocation that can be easily avoided, so why not?

> As for FrozenDictionary, I'm not interested in using it, mainly since it's not available for .NET Standard 2.0

But if you don't intend to support AOT before net8.0 it shouldn't be an issue, right?

> Further, [tests show that it's slower for very small dictionaries](https://code-corner.dev/2023/11/08/NET-8-%E2%80%94-FrozenDictionary-performance/)

Didn't know that. Noted.
